SonarQube.Common/SarifUtilities.cs

```diff
@@ -0,0 +1,91 @@
+//   Copyright (c) SonarSource SA and Microsoft Corporation.  All rights reserved.
+//   Licensed under the MIT License. See License.txt in the project root for license information.
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace SonarQube.Common
+{
+    public class SarifUtilities
+    {
+        /// <summary>
+        /// Returns true if the given SARIF came from the VS 2015 RTM Roslyn, which does not provide correct output.
+        /// </summary>
+        public static bool IsSarifFromUnsupportedCompiler(string input)
+        {
+            // low risk of false positives / false negatives
+            if (input.Contains(@"""toolName"": ""Microsoft (R) Visual C# Compiler""")
+                && input.Contains(@"""productVersion"": ""1.0.0"""))
+            {
+                return true;
+            }
+
+            // all other cases
+            return false;
+        }
+
+        /// <summary>
+        /// Used to correct invalid SARIF generated by VS 2015 RTM Roslyn. 
+        /// Replaces all slashes in any line beginning with "uri": with double slashes.
+        /// Applying this to already-valid SARIF will result in over-escaping.
+        /// </summary>
+        public static bool FixImproperlyEscapedSarif(string input, out string output)
+        {
+            string[] inputLines = input.Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
+            bool changeApplied = false;
+
+            if (IsSarifFromUnsupportedCompiler(input))
+            {
+                for (int i = 0; i < inputLines.Length; i++)
+                {
+                    string line = inputLines[i];
+                    if (line.Contains(@"""uri"": ")
+                        || line.Contains(@"""shortMessage"": ")
+                        || line.Contains(@"""fullMessage"": ")
+                        || line.Contains(@"""title"": "))
+                    {
+                        line = line.Replace(@"\", @"\\");
+
+                        string[] subStrings = line.Split('"');
+                        if (subStrings.Length > 5)
+                        {
+                            string[] valueStrings = new string[subStrings.Length - 4];
+                            Array.Copy(subStrings, 3, valueStrings, 0, subStrings.Length - 4);
+                            string newValue = String.Join("\\\"", valueStrings); // join value string together with escaped quotes
+
+                            string[] newLineStrings = new string[5]
+                            {
+                                subStrings[0],
+                                subStrings[1],
+                                subStrings[2],
+                                newValue,
+                                subStrings[subStrings.Length - 1]
+                            }; // construct final line
```

Comment:
Using regular expressions would have probably have been neater and more declarative.